src/main/ngapp/accio/src/app/mask.directive.ts

```diff
@@ -0,0 +1,140 @@
+import { Directive, HostListener, ElementRef, Input } from '@angular/core';
+import { MagicWandService } from './magic-wand.service';
+
+@Directive({
+  selector: '[appMask]',
+})
+export class MaskDirective {
+  @Input() scale: number;
+  @Input() originalData: ImageData;
+
+  // create direct reference of canvas on editor.html
+  constructor(
+    private canvas: ElementRef<HTMLCanvasElement>, 
+    private magicWandService: MagicWandService,
+  ) { }
+  private ctx: CanvasRenderingContext2D;
+  private maskImage: HTMLImageElement;
+
+  @HostListener('click', ['$event'])
+  onClick(e: MouseEvent) {
+    const xCoord = e.offsetX;
+    const yCoord = e.offsetY;
+    this.ctx = this.canvas.nativeElement.getContext('2d');
+    let imgData = this.ctx.getImageData(0,0, this.canvas.nativeElement.width, this.canvas.nativeElement.height);
+
+    // tolerance will come from user input
+    const tolerance = 20;
+
+    // returns an array indicies of each pixel in the mask
+    const maskPixels = this.magicWandService.floodfill(this.originalData, Math.floor(xCoord / this.scale), Math.floor(yCoord / this.scale), tolerance);
```

Comment:
I want to discuss our implementation here. If I understand correctly, you're using the original image as the source data here. That means when the user floodfills, a mask is created, and if they floodfill again, the floodfill algorithm will completely ignore the freshly created mask of from the previous floodfill command. We should talk more thoroughly about how adding to the most current version of the mask will work, to ensure that the user has an intuitive experience. Ie is it advantageous/convenient for the user to be able to floodfill once, then floodfill again, but have us account for the most current mask as part of the image being floodfilled.

Reply:
Yes, I agree. I thought when we talked about the undo/redo, the plan was to have the magic wand api store the cumulative set of all pixels that should be in the mask. For example, if, on the image of four colors, the user clicks the top left quad, then those pixels are added to the set of pixels and saved in the master queue or list. Then when the user clicks the top right quad, the new mask will be created for that quad and we can combine the set created before with the new set created then store the combined set, and pass it back to this directive. You're right that currently the image is just the newest mask, but it appears as if the whole image is in the mask because I get the context of the image displayed when the user clicks on the image, which the old mask is technically still on, once we get the master list/queue of all sets created and combined I will change this.ctx to be a reference to the scaled image in editor.component.ts. I probably should have added comments to explain this and will do that now. Let me know what you're thinking/if that makes sense.

Reply:
Alternatively, or in addition, I can create the mask image in editor.component, create the Imagedata for it and then pass the data through so that each time a pixel is added we don't have to redraw the entire mask. I would say in addition though because we will still need the queue/set for undo/redo.

Reply:
So after we implement the idea you're referring to, what would be the result of implementing floodfill on this example:

PREMISE:
There is an image with 3 colors--A, B, and C--oriented like a stack. A is adjacent to B is adjacent to C. Like so:
A
B
C
Floodfill has already been called once by clicking on a B pixel, so mask M would cover the image and the user would see:
A
M
C

PROBLEM:
Now supposed C is a color such that the tolerance would allow floodfill to capture B as well as A inside a mask if we use C as the basis for floodfill's tolerance logic. 
If we clicked on C to do another floodfill command, would the result be:
1.
M
M
M
or 2.
A
M
M

Reply:
2 would be the result to my knowledge. Although that is not what I thought you were talking about. So are what you are saying is you want the mask to be 1 not 2?

Reply:
No, outcome 2 makes more sense to me. LGTM

Reply:
Just to clarify, to get result 1 implies the floodfill can potentially run over all previously masked pixels every time, correct?

Reply:
@dtjanaka yes, and so I wanted to make sure to distinguish that possibility and the difference

Reply:
A couple comments.

- I really liked be able to add to mask with multiple clicks
- I think having the default be to add and then a button to clear the mask would be nice
- Adding all these details in this thread to the design document would be great if they aren't already there.

Reply:
Something I just realized.  In order to get 2, this means that the floodfill needs to know to stop at previously masked pixels.  I think that seems to work in the current implementation because the clicked on color is not in tolerance of black.  If it was then it would add the masked pixels into the mask.

Reply:
I added these to the design doc under UI and alternative ideas. When you say to clear the mask, do you mean remove the entire mask the user just created? I think that makes sense to have I just wanted to check.

Reply:
@robotsaurus I think the current call to floodfill uses the original ImageData for reference and is unaware of any changes to the working canvas/ImageData, and so it wouldn't "detect" the black. However it is an easy fix, we would just need to pass in a different argument for the floodfill's imgData param

Reply:
@charmychancey If we pass in the original data then it's unclear to me how we would get 2 instead of 1 under the current implementation.  Were you suggesting to modify the floodfill algorithm to achieve 2?

Reply:
@shmcaffrey Yes I was referring to clear the entire mask.  So start over.  I guess this could be achieved by refreshing the page so probably not a huge priority to implement.